azext_iot/tests/iothub/message_endpoint/conftest.py

```diff
@@ -0,0 +1,464 @@
+from typing import Optional, TypeVar
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+SubRequest = TypeVar('SubRequest')
+Mark = TypeVar('Mark')
+cli = EmbeddedCLI()
+REQUIRED_TEST_ENV_VARS = ["azext_iot_testrg"]
+settings = DynamoSettings(req_env_set=REQUIRED_TEST_ENV_VARS)
+RG = settings.env.azext_iot_testrg
+
+
+def generate_hub_id() -> str:
+    return f"test-hub-{generate_generic_id()}"[:40]
+
+
+def generate_hub_depenency_id() -> str:
+    return f"testhubdep{generate_generic_id()}"[:24]
+
+
+def get_closest_marker(request: SubRequest) -> Mark:
+    for item in request.session.items:
+        if item.get_closest_marker("hub_infrastructure"):
+            return item.get_closest_marker("hub_infrastructure")
+    return request.node.get_closest_marker("hub_infrastructure")
+
+
+def tags_to_dict(tags: str) -> dict:
+    result = {}
+    split_tags = tags.split()
+    for tag in split_tags:
+        kvp = tag.split("=")
+        result[kvp[0]] = kvp[1]
+    return result
+
+
+@pytest.fixture(scope="module")
+def provisioned_iot_hub_module(request, provisioned_user_identity_module) -> Optional[dict]:
+    result = _iot_hub_provisioner(request, provisioned_user_identity_module)
+    yield result
+    if result:
+        _iot_hub_removal(result["name"])
+
+
+@pytest.fixture(scope="module")
+def provisioned_only_iot_hub_module(request) -> Optional[dict]:
+    result = _iot_hub_provisioner(request)
+    yield result
+    if result:
+        _iot_hub_removal(result["name"])
+
+
+def _iot_hub_provisioner(request, provisioned_user_identity=None):
+    name = generate_hub_id()
+    base_create_command = f"iot hub create -n {name} -g {RG} --sku S1"
+
+    hub_marker = get_closest_marker(request)
+    desired_location = None
+    desired_tags = None
+    desired_sys_identity = False
+    desired_user_identity = False
+
+    if hub_marker:
+        desired_location = hub_marker.kwargs.get("location")
+        desired_tags = hub_marker.kwargs.get("desired_tags")
+        desired_sys_identity = hub_marker.kwargs.get("sys_identity")
+        desired_user_identity = hub_marker.kwargs.get("user_identity")
+
+    if desired_sys_identity:
+        base_create_command += " --mi-system-assigned"
+    if desired_user_identity and provisioned_user_identity:
+        user_identity_id = provisioned_user_identity["id"]
+        base_create_command += f" --mi-user-assigned {user_identity_id}"
+    if desired_tags:
+        base_create_command = base_create_command + f" --tags {desired_tags}"
+    if desired_location:
+        base_create_command = base_create_command + f" -l {desired_location}"
+
+    return cli.invoke(base_create_command).as_json()
+
+
+def _iot_hub_removal(name):
+    delete_result = cli.invoke(f"iot hub delete -n {name} -g {RG}")
+    if not delete_result.success():
+        logger.error(f"Failed to delete iot hub resource {name}.")
+
+
+@pytest.fixture(scope="module")
+def provisioned_user_identity_module() -> Optional[dict]:
```

Comment:
I made everything module but we may need fixtures with smaller scopes.

also add comments to denote sections/specific services

Reply:
added comments. Will keep as module as its prob the smallest scope that makes sense